pkg/resource/deploy/step.go

```diff
@@ -821,38 +949,87 @@ func (s *UpdateStep) Apply() (resource.Status, StepCompleteFunc, error) {
 		}
 
 		// Update to the combination of the old "all" state, but overwritten with new inputs.
-		resp, upderr := prov.Update(context.TODO(), plugin.UpdateRequest{
-			URN:                   s.URN(),
-			Name:                  s.URN().Name(),
-			Type:                  s.URN().Type(),
-			ID:                    s.old.ID,
-			OldInputs:             s.old.Inputs,
-			OldOutputs:            s.old.Outputs,
-			NewInputs:             s.new.Inputs,
-			Timeout:               s.new.CustomTimeouts.Update,
-			IgnoreChanges:         s.ignoreChanges,
-			Preview:               s.deployment.opts.DryRun,
-			ResourceStatusAddress: resourceStatusAddress,
-			ResourceStatusToken:   resourceStatusToken,
-			OldViews:              s.oldViews,
-		})
+		var resp plugin.UpdateResponse
+		failures := []string{}
+
+		for {
+			resp, err = prov.Update(context.TODO(), plugin.UpdateRequest{
+				URN:                   s.URN(),
+				Name:                  s.URN().Name(),
+				Type:                  s.URN().Type(),
+				ID:                    s.old.ID,
+				OldInputs:             s.old.Inputs,
+				OldOutputs:            s.old.Outputs,
+				NewInputs:             s.new.Inputs,
+				Timeout:               s.new.CustomTimeouts.Update,
+				IgnoreChanges:         s.ignoreChanges,
+				Preview:               s.deployment.opts.DryRun,
+				ResourceStatusAddress: resourceStatusAddress,
+				ResourceStatusToken:   resourceStatusToken,
+				OldViews:              s.oldViews,
+			})
+
+			if err == nil {
+				resourceError = nil
+				resourceStatus = resource.StatusOK
 
-		s.new.Lock.Lock()
-		defer s.new.Lock.Unlock()
+				break
+			}
 
-		if upderr != nil {
+			failures = append([]string{err.Error()}, failures...)
 			if resp.Status != resource.StatusPartialFailure {
-				return resp.Status, nil, upderr
+				return resp.Status, nil, err
 			}
 
-			resourceError = upderr
+			resourceError = err
 			resourceStatus = resp.Status
 
-			if initErr, isInitErr := upderr.(*plugin.InitError); isInitErr {
+			if initErr, isInitErr := err.(*plugin.InitError); isInitErr {
 				s.new.InitErrors = initErr.Reasons
```

Comment:
This was previously running inside `s.new.Lock.Lock()`. Looking at the PR that added the lock https://github.com/pulumi/pulumi/pull/16148/changes#diff-e47de4c653b38357c27e7f94b1cdcce7169333b1731a5c3a99e57dbfd708cb19 I *think* we only need this lock when mutation the `State.Outputs,` not the `InitErrors` property, so this should be fine.